docs/proposed/psa-conditional-inclusion-c.md

```diff
@@ -5,7 +5,7 @@ This document is a proposed interface for deciding at build time which cryptogra
 
 This is currently a proposal for Mbed TLS. It is not currently on track for standardization in PSA.
 
-Time-stamp: "2020/09/07 08:27:32 GMT"
+Time-stamp: "2020/09/21 18:07:09 GMT"
```

Comment:
> Does it mean that if a PSA driver provides a crypto mechanism then "automatically" the Mbed TLS code providing the mechanism is removed from the Mbed TLS build? If yes, what about not to do that, at least to start with, to simplify.

The short-term goal is to make it _possible_ to include e.g. PSA RSA support without including `MBEDTLS_RSA_C` when there's a PSA transparent driver for RSA. I don't see a simplification that would make it possible but not automatic. At least not without making a temporary mechanism that we would not want to maintain backward compatibility with because it would be too ad hoc and hard to use.

> Why not just tightening mbedtls/config.h to psa/config.h in that case: the defines in psa/config.h define the defines in mbedtls/config.h?

I don't understand what you're proposing. Is this about the extra layer of indirection with `MBEDTLS_HAVE_BUILTIN_xxx` being aliases for `MBEDTLS_xxx_C`? That requires more work in `config_psa.h` but less in `psa_crypto.c` and other places. It makes the code more readable because there isn't always a 1-to-1 correspondence between `MBEDTLS_HAVE_BUILTIN_xxx` and the legacy symbols, and this the defining expression is only present once.

Reply:
> The short-term goal is to make it possible to include e.g. PSA RSA support without including MBEDTLS_RSA_C when there's a PSA transparent driver for RSA. I don't see a simplification that would make it possible but not automatic. At least not without making a temporary mechanism that we would not want to maintain backward compatibility with because it would be too ad hoc and hard to use.

My understanding of the current proposal is that, the integrator defines in psa/crypto_config.h the needs of the application, defines the drivers and their capabilities and from that the build system computes the needs for built-in crypto.

What I am suggesting is rather for the integrator to defines in psa/crypto_config.h the needs in built-in crypto. Thus the integrator has to determine herself the needs for built-in crypto given the application needs and the drivers capabilities.

To follow-up on the RSA example. if currently we disable MBEDTLS_RSA_C in config.h what do have to change in psa_crypto.c for the driver wrapper to be called for RSA operations ?

Reply:
Requiring that the integrator define the needs for built-in crypto requires the integrator to know the needs of the application and to understand the MBEDTLS configuration symbols. This puts a significant burden on the integrator which can be avoided by a small amount of code: we can write the rules to deduce the MBEDTLS symbols from the PSA ones once and for all. We need to do the analysis to figure out the requirements in order to write the rules for the other direction (which are necessary for backward compatibility), so the cost for us to write the rules for the PSA→MBEDTLS is just the cost of writing those few extra lines.

For RSA, `defined(MBEDTLS_RSA_C)` in `psa_crypto.c` will change to one of, or a combination of, `MBEDTLS_PSA_BUILTIN_KEY_TYPE_RSA`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_SIGN_RAW`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PSS`, `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_CRYPT` or `MBEDTLS_PSA_BUILTIN_ALG_RSA_OAEP` depending on the location (PSA has more granularity there). `defined(MBEDTLS_PKCS1_V15)` will change to one of `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_SIGN_RAW` or `MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_CRYPT`. `defined(MBEDTLS_PKCS1_V21)` will change to one of `MBEDTLS_PSA_BUILTIN_ALG_RSA_PSS`  or `MBEDTLS_PSA_BUILTIN_ALG_RSA_OAEP`. `defined(MBEDTLS_RSA_C) && defined(MBEDTLS_RSA_GENPRIME)` will change to will change to `MBEDTLS_PSA_BUILTIN_KEY_TYPE_RSA` (PSA has less granularity there).

Reply:
Yes I understand now that the conversion from PSA symbols to MBEDTLS symbols is not a big work and as it eases integration it makes sense to do it.

Otherwise regarding:

>> Why not just tightening mbedtls/config.h to psa/config.h in that case: the defines in psa/config.h define the defines in mbedtls/config.h?

> I don't understand what you're proposing. Is this about the extra layer of indirection with MBEDTLS_HAVE_BUILTIN_xxx being aliases for MBEDTLS_xxx_C? That requires more work in config_psa.h but less in psa_crypto.c and other places. It makes the code more readable because there isn't always a 1-to-1 correspondence between MBEDTLS_HAVE_BUILTIN_xxx and the legacy symbols, and this the defining expression is only present once.

I am not sure why when `MBEDTLS_PSA_CRYPTO_CONFIG` is set the MBEDTLS symbols are not just deduced from the PSA ones. Following-up on the AES example in the "PSA Crypto configuration file" section, why do we give the possibility of defining MBEDTLS_AES_C if given psa/crypto_config.h there is no need for it?

Reply:
When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, only the `MBEDTLS_xxx` symbols that are needed for PSA calls are automatically enabled. The PSA crypto configuration can't determine whether to enable `MBEDTLS_CIPHER_C` or `MBEDTLS_SSL_CLI_C`.

It's `psa/crypto_config.h` and not `psa/config.h` because this configuration mechanism is specific to crypto. Other PSA components manage their configuration differently (likely via JSON).

I'd forgotten to make the editorial corrections from this comment thread, I just pushed them.

Reply:
> When `MBEDTLS_PSA_CRYPTO_CONFIG` is set, only the `MBEDTLS_xxx` symbols that are needed for PSA calls are automatically enabled. The PSA crypto configuration can't determine whether to enable `MBEDTLS_CIPHER_C` or `MBEDTLS_SSL_CLI_C`.

Coming back to this, I understand your point for `MBEDTLS_SSL_CLI_C` (TLS, not crypto configuration) but I don't really understand why we would need to enable `MBEDTLS_CIPHER_C` if it is not required by `PSA_WANT_xxx` macros. Do we want to support the case when MBEDTLS_PSA_CRYPTO_CONFIG is defined where the library crypto needs are not completely defined by `PSA_WANT_xxx` macros?

Reply:
> Do we want to support the case when `MBEDTLS_PSA_CRYPTO_CONFIG` is defined where the library crypto needs are not completely defined by `PSA_WANT_xxx` macros?

Yes. A classic crypto feature may need to be enabled either because it's needed internally to implement a PSA crypto feature or because the application needs that classic feature. ``MBEDTLS_PSA_CRYPTO_CONFIG` only determines whether PSA crypto features are deduced from classic crypto options (off) or configured separately (on). If `MBEDTLS_PSA_CRYPTO_CONFIG` is enabled, the PSA crypto configuration (including driver availability) determines a set of classic crypto options that are enabled in addition to those that are enabled explicitly.

Reply:
ok, thanks for clarifying this.